core-tests/shared/src/test/scala/zio/ZLayerSpec.scala

```diff
@@ -181,6 +181,44 @@ object ZLayerSpec extends ZIOBaseSpec {
         _      <- env.use_(ZIO.unit)
         actual <- ref.get
       } yield assert(actual)(equalTo(expected))
+    },
+    testM("map does not interfere with sharing") {
```

Comment:
Good catch. Maybe there's something we can do to make new combinators more "obviously correct" with respect to sharing??

Reply:
Yes I was thinking about that too. I think like the implementation of `ZManaged` in terms of `Reservation` the underlying implementation in terms of the `MemoMap` has some unavoidable complexity so we want to get to a point where you can do almost anything in terms of a few basic combinators that make sure the right thing happens. Then the message is "use the combinators unless you really know what you are doing".

You can see that working in terms of methods like `map`, `update` and `set` that are all implemented directly or indirectly in terms of `>>>`. The problem for implementing `mapError` is that we can't change the error type with `>>>` or `++`.

I think we might want to think about getting two two basic combinators, a more powerful version of `>>>` that lets you sequentially combine two layers getting access to either the error or success of the first layer, kind of like `foldM`, and then a parallel combinator like `++` or `+!+` that lets you combine two layers in parallel with a function. I think then you could implement everything else in terms of those and have it be "right".

What do you think?

Reply:
I like that idea. I also think we should hide the constructor so people can _only_ use the provided methods. I'd rather users come to us and say, "Please add a new operator" then take the chance they get the memo map & finalizers right.

Reply:
I made the constructor private when I implemented the new encoding so it is only us who can get it wrong.